x-pack/legacy/plugins/lens/public/datatable_visualization_plugin/visualization.tsx

```diff
@@ -6,146 +6,83 @@
-                  defaultMessage: 'Title',
-                })}
-              />
-            </EuiFormRow>
+export interface DatatableVisualizationState {
+  layers: LayerState[];
+}
 
-            <EuiFormRow>
-              <EuiFlexGroup>
-                <EuiFlexItem grow={true}>
-                  <NativeRenderer
-                    data-test-subj="lnsDatatable_dimensionPanel"
-                    render={datasource.renderDimensionPanel}
-                    nativeProps={{
-                      columnId: id,
-                      dragDropContext: props.dragDropContext,
-                      filterOperations: () => true,
-                    }}
-                  />
-                </EuiFlexItem>
+function newLayerState(layerId: string): LayerState {
+  return {
+    layerId,
+    columns: [generateId()],
+  };
+}
 
-                <EuiFlexItem grow={false}>
-                  <EuiButtonIcon
-                    size="s"
-                    color="warning"
-                    data-test-subj={`lnsDatatable_dimensionPanelRemove_${id}`}
-                    iconType="trash"
-                    onClick={() => {
-                      datasource.removeColumnInTableSpec(id);
-                      const newColumns = [...state.columns];
-                      newColumns.splice(index, 1);
-                      setState({
-                        ...state,
-                        columns: newColumns,
-                      });
-                    }}
-                    aria-label={i18n.translate('xpack.lens.datatable.removeColumnAriaLabel', {
-                      defaultMessage: 'Remove',
-                    })}
-                  />
-                </EuiFlexItem>
-              </EuiFlexGroup>
-            </EuiFormRow>
-          </>
-        );
-      })}
+function updateColumns(
+  state: DatatableVisualizationState,
+  layer: LayerState,
+  fn: (columns: string[]) => string[]
+) {
+  const columns = fn(layer.columns);
+  const updatedLayer = { ...layer, columns };
+  const layers = state.layers.map(l => (l.layerId === layer.layerId ? updatedLayer : l));
+  return { ...state, layers };
+}
 
-      <div>
-        <EuiButton
-          data-test-subj="lnsDatatable_dimensionPanel_add"
-          onClick={() => {
-            const newColumns = [...state.columns];
-            newColumns.push({
-              id: generateId(),
-              label: '',
-            });
-            setState({
-              ...state,
-              columns: newColumns,
-            });
-          }}
-          iconType="plusInCircle"
-        />
-      </div>
-    </EuiForm>
+export function DataTableLayer({
+  layer,
+  frame,
+  state,
+  setState,
+  dragDropContext,
+}: { layer: LayerState } & VisualizationProps<DatatableVisualizationState>) {
+  const datasource = frame.datasourceLayers[layer.layerId];
+  return (
+    <EuiFormRow
+      label={i18n.translate('xpack.lens.datatable.columns', { defaultMessage: 'Columns' })}
+    >
```

Comment:
The table visualization isn't currently rendering the `layerPanel` provided by the datasource, which is the thing that indicates which indexpattern was being used. I think it would be helpful for consistency there.

Reply:
True. In fact, it's probably worth having a `<LayerPanel ...` component that our visualizations can all share, since the concept should render consistently across our visualizations.